torch/csrc/Module.cpp

```diff
@@ -348,9 +348,41 @@ PyObject* THPModule_swap_tensor_impl(PyObject* _unused, PyObject* args) {
 
   // Swap the Tensor Impl
   c10::MaybeOwned<at::Tensor> tmp = a->cdata;
+
+  // The TensorImpls contain PyObjectSlots that have a reference to the PyObject
+  // associated with the TensorImpl. Swap this field as well.
+  c10::optional<PyObject*> mb_obj_a =
```

Comment:
You could just use `a_` here. that should be the same object.
If we want to be extra safe, you can keep this and add another assert that mb_obj_a == a_.